host/s3.go

```diff
@@ -0,0 +1,412 @@
+	return
+}
+
+// s3KeySubmission is a unitary method (mostly for testing) that submits any arbitrary []byte into S3
+func (s3ds *S3DataStream) s3KeySubmission(ctx context.Context, s3Key string, content []byte) error {
+	slog.Debug(
+		"submitting traces to s3",
+		slog.Attr{Key: "file", Value: slog.StringValue(s3Key)},
+		slog.Attr{Key: "s3-bucket", Value: slog.StringValue(s3ds.config.Bucket)},
+	)
+	// get the file descriptor for the s3 file
+	s3OpCtx, cancel := context.WithTimeout(ctx, S3OpTimeout)
+	defer cancel()
+
+	_, err := s3ds.client.PutObject(s3OpCtx, &s3.PutObjectInput{
+		Bucket: aws.String(s3ds.config.Bucket),
+		Key:    aws.String(s3Key),
+		Body:   bytes.NewReader(content),
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getObjsInBucket returns all the existing items in the s3 bucket
+func (s3ds *S3DataStream) getObjsInBucket(ctx context.Context) ([]types.Object, error) {
+	slog.Debug(
+		"listing items on s3 bucket",
+		slog.Attr{Key: "bucket-name", Value: slog.StringValue(s3ds.config.Bucket)},
+	)
+	opCtx, cancel := context.WithTimeout(ctx, S3OpTimeout)
+	defer cancel()
+
+	objects, err := s3ds.client.ListObjectsV2(opCtx, &s3.ListObjectsV2Input{
+		Bucket: &s3ds.config.Bucket,
+	})
+	if err != nil {
+		return []types.Object{}, err
+	}
+	return objects.Contents, nil
+}
+
+// listBuckets returns all the available buckets in the given s3 instance
+func (s3ds *S3DataStream) listBuckets(ctx context.Context) ([]types.Bucket, error) {
+	slog.Debug("listing s3 buckets")
+	opCtx, cancel := context.WithTimeout(ctx, S3OpTimeout)
+	defer cancel()
+
+	buckets, err := s3ds.client.ListBuckets(opCtx, &s3.ListBucketsInput{})
+	if err != nil {
+		return []types.Bucket{}, err
+	}
+	return buckets.Buckets, nil
+}
+
+// testConnection checks if there is any available connection with the s3 instance
+// then lists the exising buckets.
+// returns error if it can't perform the operation or if the given bucket isn't present
+func (s3ds *S3DataStream) testConnection(ctx context.Context) error {
+	slog.Debug("testing s3 connection")
+	opCtx, cancel := context.WithTimeout(ctx, S3OpTimeout)
+	defer cancel()
+
+	buckets, err := s3ds.listBuckets(opCtx)
+	if err != nil {
+		return err
+	}
+
+	for _, bucket := range buckets {
+		if *bucket.Name == s3ds.config.Bucket {
+			slog.Info(
+				"successfull connection to the S3 bucket",
+				slog.Attr{Key: "bucket", Value: slog.StringValue(*bucket.Name)},
+			)
+			return nil
+		}
+	}
+	return fmt.Errorf("Couldn't find bucket %s among existing ones in the s3 instance", s3ds.config.Bucket)
+}
+
+// removeItemFromS3 removes the item from the s3 instance (for testing purposes)
+func (s3ds *S3DataStream) removeItemFromS3(ctx context.Context, s3Key string) error {
+	slog.Debug(
+		"launching S3 periodic flusher",
+		slog.Attr{Key: "s3key", Value: slog.StringValue(s3Key)},
+	)
+	opCtx, cancel := context.WithTimeout(ctx, S3OpTimeout)
+	defer cancel()
+
+	s3ds.client.DeleteObject(opCtx, &s3.DeleteObjectInput{
```

Comment:
doesn't this return an error?

Reply:
ah just for testing. Maybe still worth adding if at some point in the future someone just uses this method and expects to see an error?

Reply:
100% right